[bmalloc] Add fuzzing option for TZone compaction

Achierius opened the pull request:
#### edcd534cc66fc2e5037f85a5acd120cd9b5d1f76
<pre>
[bmalloc] Add fuzzing option for TZone compaction
<a href="https://bugs.webkit.org/show_bug.cgi?id=295807">https://bugs.webkit.org/show_bug.cgi?id=295807</a>
<a href="https://rdar.apple.com/155639885">rdar://155639885</a>

Reviewed by Mark Lam.

This allows us to test different configurations for the compaction of
types allocated with TZoneMalloc.

Currently this is only compiled-in in Debug builds, as the extra runtime
check it requires on the malloc hot-path would otherwise be pessimizing.
When it is compiled in, it is enabled with the envvar
&apos;bmalloc_DynamicCompactModeEnable=1&apos;.
To change the seed used from the default (which is tied to the TZone
seed), you can also set &apos;bmalloc_DynamicCompactModeSeed=0x&lt;seed&gt;&apos;.

* Source/bmalloc/bmalloc/BPlatform.h:
* Source/bmalloc/bmalloc/TZoneHeap.cpp:
(bmalloc::api::getTZoneDynamicCompactionSalt):
(bmalloc::api::isTZoneDynamicCompactionEnabled):
* Source/bmalloc/bmalloc/TZoneHeap.h:
(bmalloc::api::fnv1aHash):
(bmalloc::api::encodeTZoneDynamicCompactionKey):
(bmalloc::api::shouldDynamicallyCompact):
(bmalloc::api::requiresCompactPointers):
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
(bmalloc::api::TZoneHeapManager::dynamicCompactionSalt):
(bmalloc::api::TZoneHeapManager::init):
(bmalloc::api::TZoneHeapManager::initializeTZoneDynamicCompaction):
(bmalloc::api::TZoneHeapManager::TZoneHeapManager::heapRefForTZoneTypeDifferentSize):
* Source/bmalloc/bmalloc/TZoneHeapManager.h:
(bmalloc::api::TZoneHeapManager::tzoneDynamicCompactionEnabled):

Canonical link: <a href="https://commits.webkit.org/298007@main">https://commits.webkit.org/298007@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46f77ec0c7dd02c61898f632bbfff59e9eeb2fcb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113888 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33640 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24099 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120052 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/64672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34268 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42201 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/86561 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/64672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116836 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27278 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102294 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66934 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/26479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20424 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63771 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/106335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96646 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20540 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123284 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/112451 "Built successfully and passed tests") | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40933 "Failed to checkout and rebase branch from PR 47911") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/30478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41306 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98506 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/95171 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24269 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40309 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/18115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40802 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136657 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/40440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36598 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/43740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/42198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.h:
```diff
@@ -166,11 +221,15 @@ public: \
         BASSERT(::bmalloc::api::tzoneMallocFallback > TZoneMallocFallback::ForceDebugMalloc); \
         if constexpr (::bmalloc::api::requiresCompactPointers<_type>()) \
             return ::bmalloc::api::tzoneAllocateCompact(s_heapRef); \
+        if (::bmalloc::api::shouldDynamicallyCompact(s_heapSpec)) \
+            return ::bmalloc::api::tzoneAllocateCompact(s_heapRef); \
```
`shouldDynamicallyCompact()` is only defined when `BUSE_DYNAMIC_TZONE_COMPACTION`.  This needs to be stubbed out with a constexpr always false function in order for `! BUSE_DYNAMIC_TZONE_COMPACTION` builds to not be affected.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.h:
```diff
@@ -122,8 +128,45 @@ struct TZoneSpecification {
+    uint64_t signature = __builtin_popcountll(key & getTZoneDynamicCompactionSalt());
+    return signature & 0x1;
+}
+
```
Why export 2 functions (getTZoneDynamicCompactionSalt and `isTZoneDynamicCompactionEnabled`) when you can export just 1 (`shouldDynamicallyCompact`).

1. Let's make `shouldDynamicallyCompact` non-inline and export it.  
2. Then get rid of `isTZoneDynamicCompactionEnabled` and `getTZoneDynamicCompactionSalt` altogether.  Instead, just call the underlying `TZoneHeapManager` methods.
3. Make `TZoneHeapManager::dynamicCompactionSalt()` `BINLINE`.  Do away with the 2 RELEASE_BASSERTs in there.  The only way that we can get to the code path where this function is called is it the type's `s_heapRef` has been initialized i.e. the slow path alloc has been called.  You can RELEASE_BASSERT in the slow path that `s_state >= State::Seeded` (I suspect that that is already asserted), and that if `s_tzoneDynamicCompactionEnabled`, then `TZoneHeapManager:: m_tzoneDynamicCompactionSalt` better be already set to a non-zero value.  You can also ensure on initialization that it is non-zero?

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.h:
```diff
@@ -132,6 +175,18 @@ inline constexpr bool requiresCompactPointers()
+    if constexpr (requires { std::remove_pointer_t<T>::allowCompactPointers; })
+        return std::remove_pointer_t<T>::allowCompactPointers;
+    return false;
+}
```
This looks like old code.  `requiresCompactPointers()` does not need to be differentiated based on `BUSE_DYNAMIC_TZONE_COMPACTION`.  Please fix.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeapInlines.h:
```diff
@@ -90,15 +92,22 @@ private: \
+{ \
+    return ::bmalloc::api::shouldDynamicallyCompact(s_heapSpec); \
+} \
+\
```
I think this code is never called.  Please remove.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.h:
```diff
@@ -188,15 +247,18 @@ public: \
 private: \
-    static _exportMacro BNO_INLINE void* operatorNewSlow(size_t);
+    static _exportMacro BNO_INLINE void* operatorNewSlow(size_t); \
+    static _exportMacro bool shouldDynamicallyCompact();
```
Not used.  Please remove.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
```diff
@@ -513,6 +522,23 @@ BINLINE unsigned TZoneHeapManager::tzoneBucketForKey(const TZoneSpecification& s
 
+bool TZoneHeapManager::s_tzoneDynamicCompactionEnabled = false;
+
+void TZoneHeapManager::initializeTZoneDynamicCompaction()
```
nit: You might want to RELEASE_BASSERT in here that the seed state is as expected.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -165,17 +170,21 @@ class TZoneHeapManager {
     TZoneTypeBuckets* populateBucketsForSizeClass(LockHolder&, SizeAndAlignment::Value);
 
     static TZoneHeapManager::State s_state;
+    static bool s_tzoneDynamicCompactionEnabled;
```
Guard this with `BUSE_DYNAMIC_TZONE_COMPACTION`.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -165,17 +170,21 @@ class TZoneHeapManager {
     unsigned largestBucketCount { 0 };
     Vector<SizeAndAlignment::Value> m_typeSizes;
 #endif
+    uint64_t m_tzoneDynamicCompactionSalt;
```
Guard this with `BUSE_DYNAMIC_TZONE_COMPACTION`.  Maybe group together with `s_tzoneDynamicCompactionEnabled`.  Size is not critical here since `TZoneHeapManager` is a singleton.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -132,6 +132,11 @@ class TZoneHeapManager {
+    BINLINE bool tzoneDynamicCompactionEnabled()
+    {
+        return s_tzoneDynamicCompactionEnabled;
+    }
```
nit: For single statement accessors like this, it is OK to put this all on one line like this:
```
BINLINE bool tzoneDynamicCompactionEnabled() { return s_tzoneDynamicCompactionEnabled; }
```
I prefer single lne for these trivial accessor functions because it makes the code more compact.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -132,6 +135,10 @@ class TZoneHeapManager {
     BEXPORT static void setBucketParams(unsigned smallSizeCount, unsigned largeSizeCount = 0, unsigned smallSizeLimit = 0);
+#if BUSE_DYNAMIC_TZONE_COMPACTION
+    BINLINE uint64_t dynamicCompactionSalt() { return m_tzoneDynamicCompactionSalt; }
+    BINLINE bool tzoneDynamicCompactionEnabled() { return s_tzoneDynamicCompactionEnabled; }
```
Typo: I think you mean `g_tzoneDynamicCompactionEnabled`.  And this function should be `static`.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
```diff
@@ -513,6 +514,26 @@ BINLINE unsigned TZoneHeapManager::tzoneBucketForKey(const TZoneSpecification& s
+        m_tzoneDynamicCompactionSalt = WeakRandom::generate(m_tzoneKeySeed);
+        if constexpr (verbose)
+            TZONE_LOG_DEBUG("dynamicCompactionSalt: 0x%llx\n", m_tzoneDynamicCompactionSalt);
+        RELEASE_BASSERT(m_tzoneDynamicCompactionSalt);
```
`WeakRandom::generate` is not guaranteed to generate a non-zero value.  I suggest you put this initialization in a `while (!m_tzoneDynamicCompactionSalt)` loop, which changes the seed if `m_tzoneDynamicCompactionSalt` is 0.

Alternatively, why must it not be 0?  I was previously thinking of using that as an indicator that this feature is disabled.  But you have a distinct `g_tzoneDynamicCompactionEnabled` flag for that.  So, maybe you can just remove this `RELEASE_BASSERT`.

Scratch that.  The way you use the salt value is a & mask with the type specific key.  So, if the salt is 0, it will filter out all types.  So, ensuring this is non-zero is the right thing to do.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.h:
```diff
@@ -122,16 +128,62 @@ struct TZoneSpecification {
+extern bool g_tzoneDynamicCompactionEnabled;
+inline bool shouldDynamicallyCompact(const TZoneSpecification& spec)
+{
+    if (g_tzoneDynamicCompactionEnabled)
```
Shouldn't this be a check for `! g_tzoneDynamicCompactionEnabled`?

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.cpp:
```diff
@@ -109,6 +109,19 @@ void tzoneFree(void* p)
+bool shouldDynamicallyCompactImpl(const TZoneSpecification& spec)
+{
+    if (!TZoneHeapManager::singleton().tzoneDynamicCompactionEnabled())
+        return false;
```
This check is now redundant.  Convert into a debug ASSERT (the bmalloc one)?

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.h:
```diff
@@ -122,16 +128,62 @@ struct TZoneSpecification {
+
+#else // !BUSE_DYNAMIC_TZONE_COMPACTION
+
+inline constexpr bool shouldDynamicallyCompact([[maybe_unused]] const TZoneSpecification& spec)
```
Drop the `spec` param name since it is not used.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.h:
```diff
@@ -122,16 +128,62 @@ struct TZoneSpecification {
     return false;
 }
 
+
```
Please remove this unnecessary empty line.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.h:
```diff
@@ -122,16 +128,62 @@ struct TZoneSpecification {
+
+#endif // BUSE_DYNAMIC_TZONE_COMPACTION
+
+    template<typename T>
```
Please unindent this `template` declarator line.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeap.h:
```diff
@@ -220,7 +278,7 @@ private: \
 
 #define MAKE_BTZONE_MALLOCED_TEMPLATE_IMPL(_templateParameters, _type) \
-    _templateParameters ::bmalloc::api::HeapRef _type::s_heapRef
+    _templateParameters ::bmalloc::api::HeapRef _type::s_heapRef;
```
Please remove this semicolon at the end.  Adding it here will result in an extra semicolon in all the places that uses it.  I wonder why your build isn't already failing.

Achierius replied:
This wouldn't cause the build to fail -- an extra semicolon is essentially a no-op.

(Obviously I'll remove it, just justifying how I was able to build/run tests &c)

MenloDorian reviewed (approved):
r=me with fixes.

MenloDorian commented on Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
```diff
@@ -509,6 +510,27 @@ BINLINE unsigned TZoneHeapManager::tzoneBucketForKey(const TZoneSpecification& s
+        g_tzoneDynamicCompactionEnabled = true;
+
+        while (!m_tzoneDynamicCompactionSalt)
+            m_tzoneDynamicCompactionSalt = WeakRandom::generate(m_tzoneKeySeed);
```
Marcus, this cannot possibly work.  If `WeakRandom::generate(m_tzoneKeySeed)` produces 0, this is going to loop forever.

MenloDorian reviewed (approved):
r=me.  nit: instead of using the term "DynamicCompaction", can you use "DynamicCompactMode" (or equivalent) instead?  "DynamicCompaction" reads like this is going to apply some memory compression, but it is really about selecting Compact mode allocation.  I strongly suggest renaming it (in the code, comments, commit message, and PR title).  Thanks

Achierius committed:
[bmalloc] Add fuzzing option for TZone compaction
https://bugs.webkit.org/show_bug.cgi?id=295807
rdar://155639885

Reviewed by Mark Lam.

This allows us to test different configurations for the compaction of
types allocated with TZoneMalloc.

Currently this is only compiled-in in Debug builds, as the extra runtime
check it requires on the malloc hot-path would otherwise be pessimizing.
When it is compiled in, it is enabled with the envvar
'bmalloc_DynamicCompactModeEnable=1'.
To change the seed used from the default (which is tied to the TZone
seed), you can also set 'bmalloc_DynamicCompactModeSeed=0x<seed>'.

* Source/bmalloc/bmalloc/BPlatform.h:
* Source/bmalloc/bmalloc/TZoneHeap.cpp:
(bmalloc::api::getTZoneDynamicCompactionSalt):
(bmalloc::api::isTZoneDynamicCompactionEnabled):
* Source/bmalloc/bmalloc/TZoneHeap.h:
(bmalloc::api::fnv1aHash):
(bmalloc::api::encodeTZoneDynamicCompactionKey):
(bmalloc::api::shouldDynamicallyCompact):
(bmalloc::api::requiresCompactPointers):
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
(bmalloc::api::TZoneHeapManager::dynamicCompactionSalt):
(bmalloc::api::TZoneHeapManager::init):
(bmalloc::api::TZoneHeapManager::initializeTZoneDynamicCompaction):
(bmalloc::api::TZoneHeapManager::TZoneHeapManager::heapRefForTZoneTypeDifferentSize):
* Source/bmalloc/bmalloc/TZoneHeapManager.h:
(bmalloc::api::TZoneHeapManager::tzoneDynamicCompactionEnabled):

Canonical link: https://commits.webkit.org/298007@main
